# Expenses Data Analysis Development 💰

About a month ago, I set a spending goal for August based on expenditure data from May to July. Today, I am going to check whether I achieved the goal I set. Additionally, I am going to further develop this small project to keep track of my spending habits every month.

## 1. Check if the goal was achieved

In [2]:
# Import the required libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
# Load the CSV file
df = pd.read_csv('../data/expenses2.csv')

In [ ]:
df.head()

items   prices         categories         date
0                  daiso   ₩4,000  Daily necessities  May 1, 2025
1            olive young  ₩42,800  Daily necessities  May 1, 2025
2             sun screen  ₩21,600  Daily necessities  May 1, 2025
3  pado cardigan pattern  ₩10,000            Hobbies  May 1, 2025
4               slippers   ₩7,150  Daily necessities  May 1, 2025

In [ ]:
# Preprocess the prices column for analysis
df['prices'] = df['prices'].str.replace('₩', '')
df['prices'] = df['prices'].str.replace(',', '')
df['prices'] = df['prices'].astype('int')

# Convert the date column from object to datetime
df['date'] = pd.to_datetime(df['date'])

# Add a 'month' column to group and analyze the data by month
df['month'] = df['date'].dt.month_name()

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148 entries, 0 to 147
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   items       148 non-null    object        
 1   prices      148 non-null    int64         
 2   categories  148 non-null    object        
 3   date        148 non-null    datetime64[ns]
 4   month       148 non-null    object        
dtypes: datetime64[ns](1), int64(1), object(3)
memory usage: 5.9+ KB


In [ ]:
df.head()

items  prices         categories       date month
0                  daiso    4000  Daily necessities 2025-05-01   May
1            olive young   42800  Daily necessities 2025-05-01   May
2             sun screen   21600  Daily necessities 2025-05-01   May
3  pado cardigan pattern   10000            Hobbies 2025-05-01   May
4               slippers    7150  Daily necessities 2025-05-01   May

Here were the goals I set last month:

1. Keep total spending under KRW 500,000
2. Limit spending on hobbies to under KRW 50,000

In [ ]:
total = df.loc[df['month']=='August', 'prices'].sum()

if total > 500000:
    print(f"You've spent KRW {(total - 500000):,} over budget.\n")
    print(f'💸 total spending : KRW {total:,}')
else:
    print("Well done! You've managed to save money this month.\n")
    print(f'💸 total spending : KRW {total:,}')

Well done! You've managed to save money this month.

💸 total spending : KRW 496,086


The total spending was KRW 496,086, which is under KRW 500,000. Therefore, I can say that I achieved my goal for total spending.

In [ ]:
hobby = df.loc[(df['month']=='August') & (df['categories']=='Hobbies'), 'prices'].sum()

if hobby > 50000:
    print(f"You've spent KRW {(hobby - 50000):,} on hobbies, which is over budget.\n")
    print(f'💸 total spending : KRW {hobby:,}')
else:
    print("Well done! You've managed to save money on hobbies this month.\n")
    print(f'💸 spending on hobbies : KRW {hobby:,}')

Well done! You've managed to save money on hobbies this month.

💸 spending on hobbies : KRW 33,590


The amount I spent on the hobbies category was KRW 35,590, which is under KRW 50,000. I achieved both of my spending goals.

---

I also checked which categories I spent the most money on this month.

In [ ]:
df[df['month']=='August'].groupby('categories', as_index=False)['prices'].sum().sort_values(by='prices', ascending=False).head()

categories  prices
9  Self improvement  140266
0              Cafe  100400
2         Cosmetics   56730
5           Grocery   42000
7           Hobbies   33590

The top categories were Self-Improvement, Cafes, Cosmetics, Groceries, and Hobbies.

## 2. Define a function for expenditure analysis

It is time-consuming to check expenditures like this every month. So, I’d like to make the process simpler. I am going to define a function that prints a summary of the key information.

In [ ]:
# Reload the csv file

df = pd.read_csv('/content/drive/MyDrive/personal data analysis project/expenses2.csv')

This function is very simple. The only inputs needed are the raw DataFrame and the number of the month to analyze. When these parameters are provided, the function outputs the total spending for the month, the increase or decrease ratio compared to the previous month, and the top 5 spending categories.

**Inputs:**

1. Raw DataFrame
2. The number of the month

**Outputs:**

1. Total spending for the month
2. Increase/decrease ratio compared to the previous month
3. Top 5 spending categories for the month

In [ ]:
def expenditure_analysis(df, month):

    spending = df.copy()

    # Preprocess the DataFrame
    spending['prices'] = spending['prices'].str.replace('₩', '')
    spending['prices'] = spending['prices'].str.replace(',', '')
    spending['prices'] = spending['prices'].astype('int')
    spending['date'] = pd.to_datetime(spending['date'])
    spending['month'] = spending['date'].dt.month

    print('========= Spending Report for This Month =========\n\n')

    # Calculate total spending for the month
    this_month = spending.loc[spending['month']==month, 'prices'].sum()
    print(f"💸 You've spent KRW {this_month:,} in total this month.\n")

    # Calculate the increase/decrease ratio compared to the previous month
    last_month = spending.loc[spending['month']==(month-1), 'prices'].sum()
    change = (this_month - last_month) / last_month * 100
    if change > 0 :
        print(f"📈 You've spent {change:.2f}% more than last month.\n")
    elif change == 0 :
        print("📊 You've spent the same amount as last month.\n")
    else:
        print(f"📉 You've spent {change:.2f}% less than last month.\n")

    # Find the top 5 categories where I spent the most
    print('📊 The top 5 categories you spent the most money on are:\n')
    ctgr = spending[spending['month']==month].groupby('categories', as_index=False)['prices'].sum()
    ctgr_sorted = ctgr.sort_values(by='prices', ascending=False)
    print(ctgr_sorted.head())

Here is the description of my spending in August using this function.

In [ ]:
expenditure_analysis(df, 8)

========= Spending Report for This Month =========


💸 You've spent KRW 496,086 in total this month.

📉 You've spent -27.91% less than last month.

📊 The top 5 categories you spent the most money on are:

         categories  prices
9  Self improvement  140266
0              Cafe  100400
2         Cosmetics   56730
5           Grocery   42000
7           Hobbies   33590
